src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -0,0 +1,250 @@
+        {
+            STTx const stx = STTx{std::move(txn)};
+            txFees += ripple::calculateBaseFee(view, stx);
+        }
+        txnFees += txFees;
+    }
+
+    // Calculate the BatchSigners Fees
+    std::int32_t signerCount = tx.isFieldPresent(sfBatchSigners)
+        ? tx.getFieldArray(sfBatchSigners).size()
+        : 0;
+
+    return ((signerCount + 2) * view.fees().base) + txnFees;
+}
+
+NotTEC
+Batch::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureBatch))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto const outerAccount = ctx.tx.getAccountID(sfAccount);
+    auto const flags = ctx.tx.getFlags();
+
+    if (flags & tfBatchMask)
+    {
+        JLOG(ctx.j.trace()) << "Batch: invalid flags.";
+        return temINVALID_FLAG;
+    }
+
+    if (std::popcount(
+            flags &
+            (tfAllOrNothing | tfOnlyOne | tfUntilFailure | tfIndependent)) != 1)
+    {
+        JLOG(ctx.j.trace()) << "Batch: too many flags.";
+        return temMALFORMED;
+    }
+
+    auto const& txns = ctx.tx.getFieldArray(sfRawTransactions);
+
+    if (txns.size() == 0)
+    {
+        JLOG(ctx.j.trace()) << "Batch: txns array is empty.";
+        return temARRAY_EMPTY;
+    }
+
+    if (txns.size() > maxBatchTxCount)
+    {
+        JLOG(ctx.j.trace()) << "Batch: txns array exceeds 8 entries.";
+        return temARRAY_TOO_LARGE;
+    }
+
+    auto const& hashes = ctx.tx.getFieldV256(sfTransactionIDs);
+
+    if (hashes.size() != txns.size())
+    {
+        JLOG(ctx.j.trace()) << "Batch: hashes array size does not match txns.";
+        return temMALFORMED;
+    }
+
+    if (auto const ret = preflight2(ctx); !isTesSuccess(ret))
+        return ret;
+
+    std::set<AccountID> batchSignersSet;
+    if (ctx.tx.isFieldPresent(sfBatchSigners))
+    {
+        STArray const signers = ctx.tx.getFieldArray(sfBatchSigners);
+
+        // Check that the batch signers array is not too large.
+        if (signers.size() > 8)
+        {
+            JLOG(ctx.j.trace()) << "Batch: signers array exceeds 8 entries.";
+            return temARRAY_TOO_LARGE;
+        }
+
+        // Add the batch signers to the set.
+        for (auto const& signer : signers)
+        {
+            AccountID const innerAccount = signer.getAccountID(sfAccount);
+            if (!batchSignersSet.insert(innerAccount).second)
+            {
+                JLOG(ctx.j.trace())
+                    << "Batch: Duplicate signer found: " << innerAccount;
+                return temINVALID_BATCH;
+            }
+        }
+
+        // Check the batch signers signatures.
+        auto const requireCanonicalSig =
+            ctx.rules.enabled(featureRequireFullyCanonicalSig)
+            ? STTx::RequireFullyCanonicalSig::yes
+            : STTx::RequireFullyCanonicalSig::no;
+        auto const sigResult =
+            ctx.tx.checkBatchSign(requireCanonicalSig, ctx.rules);
```

Comment:
IMO this check should be at the end of `preflight` since checking signatures is compute-intensive.

Reply:
Actually, should be after iterating through the inner transactions, but before processing the `BatchSigners`.